staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/established_controller.go

```diff
@@ -0,0 +1,126 @@
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/util/workqueue"
+)
+
+// EstablishingController controls how and when CRD is established.
+type EstablishingController struct {
+	client *internalclientset.Clientset
+	queue  workqueue.RateLimitingInterface
+}
+
+// NewEstablishingController creates new EstablishingController.
+func NewEstablishingController(client *internalclientset.Clientset) *EstablishingController {
+	return &EstablishingController{
+		client: client,
+		queue:  workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "crdEstablishing"),
+	}
+}
+
+func (ec *EstablishingController) Run(stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer ec.queue.ShutDown()
+
+	glog.Infof("Starting EstablishingController")
+	defer glog.Infof("Shutting down EstablishingController")
+
+	// only start one worker thread since its a slow moving API
+	go wait.Until(ec.runWorker, time.Second, stopCh)
+
+	<-stopCh
+}
+
+func (ec *EstablishingController) runWorker() {
+	for ec.processNextWorkItem() {
+	}
+}
+
+// processNextWorkItem deals with one key off the queue.
+// It returns false when it's time to quit.
+func (ec *EstablishingController) processNextWorkItem() bool {
+	key, quit := ec.queue.Get()
+	if quit {
+		return false
+	}
+	defer ec.queue.Done(key)
+
+	err := ec.establishCRD(key.(string))
+	if err == nil {
+		return true
+	}
+
+	utilruntime.HandleError(fmt.Errorf("%v failed with: %v", key, err))
+	ec.queue.Add(key)
+
+	return true
+}
+
+// EstablishCRD is used to turn CRDs into the Established state.
+func (ec *EstablishingController) establishCRD(key string) error {
+	// Poll because if multiple servers try to change CRD (e.g. in HA) this can fail.
+	// That failure is normal, and so, we're trying to set it to Established again.
+	crd, err := ec.client.Apiextensions().CustomResourceDefinitions().Get(key, v1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	if !apiextensions.IsCRDConditionTrue(crd, apiextensions.Established) &&
+		apiextensions.IsCRDConditionTrue(crd, apiextensions.Installing) {
+		establishedCondition := apiextensions.CustomResourceDefinitionCondition{
+			Type:    apiextensions.Established,
+			Status:  apiextensions.ConditionTrue,
+			Reason:  "InitialNamesAccepted",
+			Message: "the initial names have been accepted",
+		}
+		installingCondition := apiextensions.CustomResourceDefinitionCondition{
+			Type:    apiextensions.Installing,
+			Status:  apiextensions.ConditionFalse,
+			Reason:  "CRDEstablished",
+			Message: "the crd have been established",
+		}
+
+		// Update storage with new CRD condition.
+		apiextensions.SetCRDCondition(crd, establishedCondition)
+		// Set CRD Installing back to false as we established it.
+		apiextensions.SetCRDCondition(crd, installingCondition)
+		// Update server with new CRD condition.
+		_, err = ec.client.Apiextensions().CustomResourceDefinitions().UpdateStatus(crd)
+		if errors.IsConflict(err) {
```

Comment:
why ignore every other error?

Reply:
Can we merge the common part of both? E.g. `apiextensions.IsCRDConditionTrue(crd, apiextensions.Installing)` is true in both cases. Moreover, only the Installing message differs.